nntrainer/src/bn_layer.cpp

```diff
@@ -83,37 +93,31 @@ sharedConstTensor BatchNormalizationLayer::forwarding(sharedConstTensor in) {
   Tensor &gamma = paramsAt(static_cast<int>(BNParams::gamma)).weight;
   Tensor &beta = paramsAt(static_cast<int>(BNParams::beta)).weight;
 
-  if (trainable) {
-    Tensor deviation;
-    input = *in;
-
-    ///< current mu */
-    Tensor cmu;
-
-    cmu = input.average(0);
+  Tensor deviation;
 
+  input = *in;
+  /// @todo change trainable #524
+  if (trainable) {
+    Tensor cmu = input.average(axes_to_reduce);
+    std::cerr << input;
+    std::cerr << cmu;
     deviation = input.subtract(cmu);
 
-    this->cvar = deviation.chain()
-                   .multiply_i(deviation)
-                   .sum(0)
-                   .multiply_i(1.0f / input_dim.batch())
-                   .add_i(epsilon)
-                   .run();
+    cvar = deviation.multiply(deviation).average(axes_to_reduce).add(epsilon);
 
-    /// @todo replace momentum paramter
-    float momentum = 0.9;
+    /// @todo replace momentum parameter to prop
+    float momentum = 0.99;
     mu.multiply_i(momentum);
     mu.add_i(cmu, 1 - momentum);
     var.multiply_i(momentum);
     var.add_i(cvar, 1 - momentum);
 
     this->x_normalized = deviation.divide(cvar.apply(sqrtFloat));
-
-    this->hidden = x_normalized.chain().multiply_i(gamma).add_i(beta).run();
+    this->hidden = x_normalized.multiply(gamma).add(beta);
   } else {
-    /// NYI
-    throw std::runtime_error("not_yet_implemented");
+    deviation = input.subtract(mu);
+    this->x_normalized = deviation.divide(var.apply(sqrtFloat));
```

Comment:
I wonder why store var than std given you have to apply sqrt everytime.

Reply:
Var is smoothed out with momentum when we need to optimize. we could save/load std instead of var to optimizer more. Let's keep it var for now since other frameworks store var as well and it is easier to match.

Reply:
then lets add a todo here to shift to std in the evaluate stage to save computation.

Reply:
Done